mime.go

```diff
@@ -1,20 +1,114 @@
 package viamrtsp
 
+/*
+#cgo pkg-config: libavutil libavcodec
+#include <libavcodec/avcodec.h>
+#include <libavutil/error.h>
+#include <libavutil/opt.h>
+#include <libavutil/dict.h>
+#include <stdlib.h>
+*/
+import "C"
+
 import (
-	"bytes"
-	"image"
-	"image/jpeg"
+	"errors"
+	"sync"
+	"unsafe"
 
 	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/logging"
 	rutils "go.viam.com/rdk/utils"
 )
 
-func encodeToJPEG(img image.Image) ([]byte, camera.ImageMetadata, error) {
-	buf := new(bytes.Buffer)
-	if err := jpeg.Encode(buf, img, nil); err != nil {
-		return nil, camera.ImageMetadata{}, err
+type mimeHandler struct {
+	logger     logging.Logger
+	jpegEnc    *C.AVCodecContext
+	currentPTS int
+	mu         sync.Mutex
+}
+
+func newMimeHandler(logger logging.Logger) *mimeHandler {
+	return &mimeHandler{
+		logger:     logger,
+		currentPTS: 0,
+	}
+}
+
+func (mh *mimeHandler) convertJPEG(frame *C.AVFrame) ([]byte, camera.ImageMetadata, error) {
+	if mh.jpegEnc == nil || frame.width != mh.jpegEnc.width || frame.height != mh.jpegEnc.height {
+		if err := mh.initJPEGEncoder(frame); err != nil {
+			return nil, camera.ImageMetadata{}, err
+		}
 	}
-	return buf.Bytes(), camera.ImageMetadata{
+	if mh.jpegEnc == nil {
+		return nil, camera.ImageMetadata{}, errors.New("failed to create encoder or destination frame")
+	}
+	// Allocate a fresh packet to prevent issues with concurrent jpeg encoding
+	pkt := C.av_packet_alloc()
+	if pkt == nil {
+		return nil, camera.ImageMetadata{}, errors.New("failed to allocate packet")
+	}
+	defer C.av_packet_free(&pkt)
+	frame.pts = C.int64_t(mh.currentPTS)
+	// If this reaches max int, it will wrap around to 0
+	mh.currentPTS++
+	res := C.avcodec_send_frame(mh.jpegEnc, frame)
+	if res < 0 {
+		return nil, camera.ImageMetadata{}, newAvError(res, "failed to send frame to MJPEG encoder")
+	}
+	res = C.avcodec_receive_packet(mh.jpegEnc, pkt)
+	if res < 0 {
+		return nil, camera.ImageMetadata{}, newAvError(res, "failed to receive packet from MJPEG encoder")
+	}
+	// There is no need to create a frame for the packet, as the packet already contains the data
+	dataGo := C.GoBytes(unsafe.Pointer(pkt.data), pkt.size)
+
+	return dataGo, camera.ImageMetadata{
 		MimeType: rutils.MimeTypeJPEG,
 	}, nil
 }
+
+func (mh *mimeHandler) initJPEGEncoder(frame *C.AVFrame) error {
+	// Lock to prevent modifying encoder while it is being used concurrently.
+	// Frame param changes are rare, so we can afford to block here.
+	mh.mu.Lock()
+	defer mh.mu.Unlock()
```

Comment:
Is there a way to test this with reconfiguring and rebuilding the camera? Either in our test suite in this module, or verifying that there is no hang or orphaned routines when we reconfigure the model, restarting the module or viam-server.

Reply:
Manually tested reconfigure and we are looking good. We have a `resource.AlwaysRebuild` flag so should be pretty safe here.